### ISSUE TITLE
Compile policies with torch.compile

### ISSUE THREAD

**Issue by AdilZouitine** [huggingface/lerobot#2061]
### **Description**
LeRobot currently supports several robotics policies (ACT, Diffusion, PI0, PI0Fast, SmolVLA, TDMPC, VQBet, SAC), but they are not optimized for `torch.compile`.
Adding `torch.compile` support would significantly improve inference and training performance by compiling computational graphs ahead of time.

**Goal**: Make all policies fully compatible with `torch.compile` without graph breaks, demonstrating measurable performance improvements.

### **What Contributors Need to Do**

1. **Run the baseline benchmark** on main branch to establish performance baseline
2. **Implement torch.compile compatibility** following the guidelines below
3. **Submit performance report** showing before/after improvements
4. **Create PR** with your changes and link it to this issue

#### **1. Run Baseline Benchmark**
The script is [here](https://gist.github.com/AdilZouitine/3574664e4cf71605986b49e9148d29ab)
```bash
# Save the benchmark script as benchmark_inference_compile_lerobot.py 
python benchmark_inference_compile_lerobot.py --policy act --device cuda --output baseline_act_report.md
```
This will likely fail with compilation errors - that's expected!

#### **2. Identify and Fix Graph Breaks**
Enable verbose debugging to see what's causing graph breaks:
```python
import torch._dynamo
torch._dynamo.config.verbose = True
torch._dynamo.config.suppress_errors = True  # Don't crash on graph breaks
```

Common issues to fix:
- **Data-dependent control flow**: Replace `if tensor.item() > 0:` with `torch.where(tensor > 0, ...)`
- **Dynamic shapes**: Ensure tensor shapes are consistent across batches
- **Python loops on tensors**: Replace with vectorized operations
- **`.item()` calls**: Keep computations in tensor form
- **In-place operations**: Be careful with operations that modify tensors in-place

#### **3. Test Your Changes**
```bash
python benchmark_inference_compile_lerobot.py --policy act --device cuda --output final_act_report.md
```

#### **4. Create Your PR**
Include:
- Your changes to make the policy compile-compatible
- Both baseline and final benchmark reports
- Brief description of what you fixed

**Comment by not-heavychevy** [comment]
@AdilZouitine Hello! I can start adding `torch.compile` support today for the easier policies - SAC, ACT, VQBet, PI0Fast, and SmolVLA.

However, Diffusion, PI0 (base), and TDMPC will likely require partial or full refactors because they currently rely on iterative Python loops (e.g., diffusion timesteps, MPC/CEM horizons), data-dependent branching (including `.item()` calls), dynamic shapes/variable sequence lengths, and some in-place state updates - all of which cause graph breaks.

**Comment by fracapuano** [comment]
Hey @not-heavychevy 👋 Thank you so much for jumping on this 🥹 I am curious about something tho: why is it that SmolVLA is classified alongside the other "simple" policies? If loops are what complicates inference, then it should totally be listed alongside diffusion and flow-matching based models (SmolVLA is a conditional flow matching model too!). Do you agree with this? Asking bc I could be missing something!

**Comment by not-heavychevy** [comment]
@fracapuano I partly agree. The CFM objective itself doesn’t determine compile difficulty - the bottlenecks are runtime control flow and shape polymorphism. In our plan, SmolVLA is in the medium bucket not because it isn’t CFM, but because the likely compile blockers are variable-length prefixes, conditional branches/masking, tensor - Python escapes (.item()), and in-place ops. That’s a different class of issues from the hard cases like diffusion/CFM solvers that use long, data-dependent iterative loops. SmolVLA’s inference path doesn’t rely on such loops, so it’s simpler to make compile-friendly.

**Comment by fracapuano** [comment]
> cases like diffusion/CFM solvers that use long, data-dependent iterative loops

Thank you for your answer! This is very interesting indeed. Are you suggesting SmolVLA inference does not use diffusion/flow-matching like loops? Asking because of how `sample_actions` is implemented, like in here👇 

https://github.com/huggingface/lerobot/blob/a6ff3cfebb0304f2c378515dd30ea06fff8f473f/src/lerobot/policies/smolvla/modeling_smolvla.py#L736-L747

**Comment by not-heavychevy** [comment]
Yes, I did miss that point. I didn't examine the implementations in enough detail. However, the main purpose of my first post was not to strictly divide policies into complex and simple ones, but to emphasize that some will be easy to add compatibility with torch.compile, while others will need to be refactored.

**Comment by fracapuano** [comment]
No problems at all! In all honesty this was much testing my own understanding of what we need to have in place to allow for compilation than challenging your answer. Thank you so much for answering. Lmk what we can do to help you take care of this in case you're still interested in carrying this thing out :)

**Comment by tc-huang** [comment]
Hello @AdilZouitine,

I've noticed the `pi0fast` policy was deprecated in PR #2203, but it's still included in the benchmark script ([benchmark_inference_compile_lerobot.py](https://gist.github.com/AdilZouitine/3574664e4cf71605986b49e9148d29ab#file-benchmark_inference_compile_lerobot-py)) and this issue's description.

**Comment by fracapuano** [comment]
Hey @tc-huang sorry for jumping in here. I think the best course of action is to consider pi0-fast deprecated! We will update the issue accordingly soon!

**Comment by not-heavychevy** [comment]
Hi @fracapuano In `benchmark_inference_compile_lerobot.py` there is no SAC policy check, but I implemented it and added the check to this script. How reasonable is this, and did I even need to make this policy optimized for torch.compile?

**Comment by fracapuano** [comment]
Hey @not-heavychevy 👋 yeah I don't think I added SAC policy to the list of models originally because of its relatively small size compared to larger models (PI0/SmolVLA-like size). Happy to review your PR adding SAC to the suite tho, just ping me as reviewer whenever you're ready :))

**Comment by not-heavychevy** [comment]
I chose SAC as the first policy to implement because of its relatively small size :). I was able to achieve a 2x speedup in training, but it did not improve inference performance.

**Comment by tommyly201** [comment]
Is there interest in doing torch.compile() for other models apart from SAC? I can pick it up

**Comment by celijose** [comment]
yes ACT model and diffusion please

**Comment by tommyly201** [comment]
@AdilZouitine @fracapuano  @not-heavychevy                                                                         
                                                                                                     
  I'd like to work on torch.compile support for **ACT** and **Diffusion Policy**.                    
                                                                                                     
  Plan:                                                                                              
  1. Start with ACT (easier - no iterative loops)                                                    
  2. Then tackle Diffusion (harder - has denoising loops)                                            
                                                                                                     
  I'll:                                                                                              
  - Run the benchmark script to establish baseline                                                   
  - Fix graph breaks (data-dependent control flow, .item() calls, etc.)                              
  - Submit separate PRs for each policy linking to this issue                                        
                                                                                                     
  Expected timeline: 3-4 weeks                                                                       
                                                                                                     
  Any guidance or concerns before I start?

**Comment by tc-huang** [comment]
Hi @tommyly201 and @celijose,

I previously tried adding torch.compile support for ACT in an earlier PR #2159, but it hasn’t been updated for a while. It might still be useful as a reference if you’d like to take a look and help update it to resolve conflicts with the current main branch.

**Comment by tommyly201** [comment]
hi @tc-huang do you want to keep working on ACT so I can do Diffusion instead? Or do you think I can address both?

**Comment by tc-huang** [comment]
Hi @tommyly201,
I would like to continue contributing to ACT over the coming weeks, mostly on weekends. Any reviews or feedback would be much appreciated.

**Comment by tommyly201** [comment]
ok I'll work on Diffusion instead and
will help however I can on your PR
